File: packages/core/src/hash.js
/**
 * RECALLER — deterministic hashing and identifiers.
 *
 * JavaScript port of recaller/core/hash.py.
 * Uses 128-bit FNV-1a (four 32-bit interleaved lanes) and mulberry32 PRNG.
 * This implementation produces byte-identical output to the Python version.
 */

/**
 * Stable JSON: object keys sorted recursively so key order never changes a hash.
 */
export function stableStringify(value) {
  if (value === null || value === undefined) return 'null'
  if (typeof value === 'boolean') return value ? 'true' : 'false'
  if (typeof value === 'number') {
    if (Number.isInteger(value)) return String(value)
    return JSON.stringify(value)
  }
  if (typeof value === 'string') return JSON.stringify(value)
  if (Array.isArray(value)) {
    const items = value.map(stableStringify)
    return `[${items.join(',')}]`
  }
  if (typeof value === 'object') {
    const keys = Object.keys(value).sort()
    const body = keys.map((k) => `${JSON.stringify(String(k))}:${stableStringify(value[k])}`)
    return `{${body.join(',')}}`
  }
  return JSON.stringify(String(value))
}

/**
 * 32-bit FNV-1a hash over unicode string code units.
 */
export function fnv1a32(s, seed) {
  let h = seed >>> 0
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i)
    h = Math.imul(h, 0x01000193) >>> 0
  }
  return h >>> 0
}

/**
 * 128-bit hex digest of any JSON-serialisable value.
 */
export function hashValue(value) {
  const s = stableStringify(value)
  const lanes = [0x811c9dc5, 0x1b873593, 0x85ebca6b, 0xc2b2ae35]
  return lanes.map((seed) => fnv1a32(s, seed).toString(16).padStart(8, '0')).join('')
}

/**
 * Short human-quotable fingerprint — first 12 hex chars.
 */
export function shortHash(value) {
  return hashValue(value).slice(0, 12)
}

/**
 * Deterministic pseudo-random generator (mulberry32).
 */
export function rng(seed) {
  let a
  if (typeof seed === 'number') {
    a = seed >>> 0
  } else {
    a = fnv1a32(String(seed), 0x9e3779b9)
  }

  return function nextRandom() {
    a = (a + 0x6d2b79f5) >>> 0
    let t = a
    t = Math.imul(t ^ (t >>> 15), t | 1) >>> 0
    t = (t + Math.imul(t ^ (t >>> 7), t | 61)) >>> 0
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

/**
 * Monotonic sortable identifier with stable prefix.
 */
export function makeId(prefix, seedParts) {
  const h = hashValue(seedParts).slice(0, 10).toUpperCase()
  return `${prefix}-${h}`
}

File: recaller/core/hash.py
"""RECALLER — deterministic hashing and identifiers.

Used to fingerprint policy documents, evidence bundles and calculation
inputs so that a replay can prove it ran against the same material.
Implementation uses 128-bit FNV-1a (four 32-bit interleaved lanes) and
mulberry32 PRNG to match the JavaScript engine byte-for-byte.
"""

import json
from typing import Any, Callable, List, Union


def stable_stringify(value: Any) -> str:
    """Stable JSON: object keys sorted recursively so key order never changes a hash."""
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        if isinstance(value, float) and value.is_integer():
            # In JS: JSON.stringify(1.0) is "1"
            return str(int(value))
        return json.dumps(value)
    if isinstance(value, str):
        return json.dumps(value)
    if isinstance(value, (list, tuple)):
        items = [stable_stringify(v) for v in value]
        return f"[{','.join(items)}]"
    if isinstance(value, dict):
        keys = sorted(k for k in value.keys() if value[k] is not None or value[k] is None)
        body = [
            f"{json.dumps(str(k))}:{stable_stringify(value[k])}"
            for k in keys
            if value[k] is not None or k in value  # keep None as null
        ]
        return f"{{{','.join(body)}}}"
    return json.dumps(str(value))


def fnv1a32(s: str, seed: int) -> int:
    """32-bit FNV-1a hash over unicode string code units."""
    h = seed & 0xFFFFFFFF
    for ch in s:
        h ^= ord(ch)
        h = (h * 0x01000193) & 0xFFFFFFFF
    return h & 0xFFFFFFFF


def hash_value(value: Any) -> str:
    """128-bit hex digest of any JSON-serialisable value."""
    s = stable_stringify(value)
    lanes = [0x811C9DC5, 0x1B873593, 0x85EBCA6B, 0xC2B2AE35]
    hex_parts = [f"{fnv1a32(s, seed):08x}" for seed in lanes]
    return "".join(hex_parts)


def short_hash(value: Any) -> str:
    """Short human-quotable fingerprint — first 12 hex chars."""
    return hash_value(value)[:12]


def rng(seed: Union[int, str]) -> Callable[[], float]:
    """Deterministic pseudo-random generator (mulberry32)."""
    if isinstance(seed, (int, float)):
        a = int(seed) & 0xFFFFFFFF
    else:
        a = fnv1a32(str(seed), 0x9E3779B9)

    def next_random() -> float:
        nonlocal a
        a = (a + 0x6D2B79F5) & 0xFFFFFFFF
        t = a
        t1 = (t ^ (t >> 15)) & 0xFFFFFFFF
        t2 = (t | 1) & 0xFFFFFFFF
        t = (t1 * t2) & 0xFFFFFFFF
        t3 = (t ^ (t >> 7)) & 0xFFFFFFFF
        t4 = (t | 61) & 0xFFFFFFFF
        t = (t + ((t3 * t4) & 0xFFFFFFFF)) & 0xFFFFFFFF
        return ((t ^ (t >> 14)) & 0xFFFFFFFF) / 4294967296.0

    return next_random


def make_id(prefix: str, seed_parts: Any) -> str:
    """Monotonic sortable identifier with stable prefix."""
    h = hash_value(seed_parts)[:10].upper()
    return f"{prefix}-{h}"
